statistics/handle/handle_hist.go

```diff
@@ -0,0 +1,359 @@
+	if hist != nil && h.updateCachedColumn(col, hist) {
+		task.ResultCh <- col
+	}
+	h.finishWorking(col)
+	return nil
+}
+
+func (h *Handle) getFreshStatsReader(readerCtx *statsReaderContext, ctx sqlexec.RestrictedSQLExecutor) {
+	if readerCtx.reader == nil || readerCtx.createdTime.Add(h.Lease()).Before(time.Now()) {
+		if readerCtx.reader != nil {
+			err := h.releaseStatsReader(readerCtx.reader, ctx)
+			if err != nil {
+				logutil.BgLogger().Warn("Fail to release stats loader: ", zap.Error(err))
+			}
+		}
+		for {
+			newReader, err := h.getStatsReader(0, ctx)
+			if err != nil {
+				logutil.BgLogger().Error("Fail to new stats loader, retry after a while.", zap.Error(err))
+				time.Sleep(time.Millisecond * 10)
+			} else {
+				readerCtx.reader = newReader
+				readerCtx.createdTime = time.Now()
+				return
+			}
+		}
+	} else {
+		return
+	}
+}
+
+// readStatsForOne reads hist for one column, TODO load data via kv-get asynchronously
+func (h *Handle) readStatsForOne(col model.TableColumnID, c *statistics.Column, reader *statsReader) (*statistics.Column, error) {
+	hg, err := h.histogramFromStorage(reader, col.TableID, c.ID, &c.Info.FieldType, c.Histogram.NDV, 0, c.LastUpdateVersion, c.NullCount, c.TotColSize, c.Correlation)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	cms, topN, err := h.cmSketchAndTopNFromStorage(reader, col.TableID, 0, col.ColumnID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	fms, err := h.fmSketchFromStorage(reader, col.TableID, 0, col.ColumnID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	rows, _, err := reader.read("select stats_ver from mysql.stats_histograms where is_index = 0 and table_id = %? and hist_id = %?", col.TableID, col.ColumnID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if len(rows) == 0 {
+		logutil.BgLogger().Error("fail to get stats version for this histogram", zap.Int64("table_id", col.TableID), zap.Int64("hist_id", col.ColumnID))
+	}
+	colHist := &statistics.Column{
+		PhysicalID: col.TableID,
+		Histogram:  *hg,
+		Info:       c.Info,
+		CMSketch:   cms,
+		TopN:       topN,
+		FMSketch:   fms,
+		IsHandle:   c.IsHandle,
+		StatsVer:   rows[0].GetInt64(0),
+	}
+	// Column.Count is calculated by Column.TotalRowCount(). Hence, we don't set Column.Count when initializing colHist.
+	colHist.Count = int64(colHist.TotalRowCount())
+	return colHist, nil
+}
+
+// drainColTask will hang until a column task can return.
+func (h *Handle) drainColTask(exit chan struct{}) (*NeededColumnTask, error) {
+	// select NeededColumnsCh firstly, if no task, then select TimeoutColumnsCh
+	for {
+		select {
+		case <-exit:
+			return nil, errExit
+		case task, ok := <-h.StatsLoad.NeededColumnsCh:
+			if !ok {
+				return nil, errors.New("drainColTask: cannot read from NeededColumnsCh, maybe the chan is closed")
+			}
+			if time.Now().After(task.ToTimeout) {
+				h.writeToTimeoutCh(task)
+				continue
+			}
```

Comment:
We could add some comments here to help others understand.

Reply:
OK, will do.

Reply:
Done.